cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -128,6 +146,55 @@ protected void shutDown() throws Exception {
     LOG.info("Stopped {}", getClass().getSimpleName());
   }
 
+  /**
+   * Scans the ProvisionerDataset for any tasks that should be in progress but are not being executed and consumes
+   * them.
+   */
+  @VisibleForTesting
+  void resumeTasks(Consumer<ProgramRunId> taskCleanup) {
+    // TODO: CDAP-13462 read tasks in chunks to avoid tx timeout
+    List<ProvisioningTaskInfo> clusterTaskInfos = getInProgressTasks();
+
+    for (ProvisioningTaskInfo provisioningTaskInfo : clusterTaskInfos) {
+      ProvisioningTaskKey taskKey = provisioningTaskInfo.getTaskKey();
+      if (taskExecutor.isRunning(taskKey)) {
+        continue;
+      }
+      ProgramRunId programRunId = provisioningTaskInfo.getProgramRunId();
+
+      ProvisioningOp provisioningOp = provisioningTaskInfo.getProvisioningOp();
+      String provisionerName = provisioningTaskInfo.getProvisionerName();
+      Provisioner provisioner = provisionerInfo.get().provisioners.get(provisionerName);
+
+      if (provisioner == null) {
+        // can happen if CDAP is shut down in the middle of a task, and a provisioner is removed
+        LOG.error("Could not provision cluster for program run {} because provisioner {} no longer exists.",
+                  programRunId, provisionerName);
+        provisionerNotifier.orphaned(programRunId);
+        Transactionals.execute(transactional, dsContext -> {
+          ProvisionerDataset provisionerDataset = ProvisionerDataset.get(dsContext, datasetFramework);
+          provisionerDataset.deleteTaskInfo(provisioningTaskInfo.getProgramRunId());
+        });
+        continue;
+      }
+
+      Runnable task = null;
+      switch (provisioningOp.getType()) {
+        case PROVISION:
+          task = createProvisionTask(provisioningTaskInfo, provisioner);
+          break;
+        case DEPROVISION:
+          task = createDeprovisionTask(provisioningTaskInfo, provisioner, taskCleanup);
+          break;
```

Comment:
Should have a `default` to catch all (probably just log and continue) to a avoid NPE in future.

Reply:
done